src/weapon/w_015.c

```diff
@@ -76,9 +166,314 @@ static void func_ptr_80170008(Entity* self) {
+            self->step++;
+            self->velocityY = FIX(-6);
+            self->ext.weapon.lifetime = 16;
+
+            if (self->flags & FLAG_UNK_800000) {
+                prim = &g_PrimBuf[self->primIndex];
+                while (prim != NULL) {
+                    prim->drawMode |= 8;
+                    prim = prim->next;
+                }
+            }
+        }
+
+        x = self->posX.i.hi;
+        y = self->posY.i.hi;
+
+        if (self->flags & FLAG_UNK_800000) {
+            prim = &g_PrimBuf[self->primIndex];
+            while (prim) {
+                if (prim->x3 == 0) {
+                    if (self->facingLeft) {
+                        prim->x0 = x - 8 -
+                                   (D_6D000_8017A6FC[prim->x1] * prim->x2 / 15);
+                    } else {
+                        prim->x0 = x + 8 +
+                                   (D_6D000_8017A6FC[prim->x1] * prim->x2 / 15);
+                    }
+
+                    prim->y0 = y + (D_6D000_8017A71C[prim->x1] * prim->x2 / 15);
+
+                    prim->x2++;
+                    if (prim->x2 > 15) {
+                        prim->x2 = 0;
+                        prim->x1 = rand() & 0xF;
+                    }
+
+                    prim->drawMode &= ~DRAW_HIDE;
+                } else {
+                    prim->x3--;
+                }
+                prim = prim->next;
+            }
+        }
+        break;
+
+    case 2:
+        self->hitboxState = 0;
+        if ((self->ext.weapon.unk7E & 7) == 0) {
+            g_api.PlaySfx(SFX_FM_EXPLODE_B);
+        }
+        self->animCurFrame = 0;
+        self->posY.val += self->velocityY;
+
+        if (--self->ext.weapon.lifetime == 0) {
+            DestroyEntity(self);
+            return;
+        }
+
+        if (self->ext.weapon.lifetime & 1) {
+            e = g_api.CreateEntFactoryFromEntity(
+                self,
+                ((g_HandId + 1) << 0xC) + 0x4E +
+                    ((self->ext.weapon.unk7E + 16) << 16),
+                0);
+            self->ext.weapon.unk7E++;
+        }
+
+        break;
+
+    case 3:
+        self->hitboxState = 0;
+        // the last three bits of weapon lifetime need to be zero.
+        if ((self->ext.weapon.lifetime & 7) == 0) {
+            g_api.PlaySfx(SFX_FM_EXPLODE_B);
+        }
+        // no, seriously. the last three bits of weapon lifetime really need
+        // to be zero. and don't trust that last comparison. i'm not joking.
+        if ((self->ext.weapon.lifetime & 7) == 0) {
+            e = g_api.CreateEntFactoryFromEntity(
+                self, ((g_HandId + 1) << 0xC) + FACTORY(0, 104), 0);
+        }
+
+        self->animCurFrame = 0;
+        if (--self->ext.weapon.lifetime == 0) {
+            DestroyEntity(self);
+            return;
+        }
+
+        break;
+    }
+}
+
+extern SpriteParts D_6D000_8017A2B0[];
+extern AnimationFrame D_6D000_8017A770[];
+extern s16 D_6D000_8017A78C[];
```

Comment:
Probably unify the externs at the top of the file

Reply:
Moved.